### ISSUE TITLE
Proposal: sub-diagnostics for finer granularity severity configuration

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#27712]
Currently, the upcoming .editorconfig support in the compiler allows diagnostic severities to be configured as follows:

```ini
dotnet_diagnostic.xUnit2017.severity = warning
```

This behavior is not flexible enough to support cases where the same diagnostic ID needs to be reported with different severity values depending on context. One example where this appears today is naming convention rules.

I propose extending the diagnostic ID form from the current `[Identifier]` to also allow `[Identifier]+[SubIdentifier]`. If a diagnostic is reported using the optional sub-identifier, the severity will be used for the sub-identifier case, if available. In all cases, the UI and compiler output would omit the sub-identifier from reporting.

```ini
# Set the default severity for naming style violations
dotnet_diagnostic.IDE1006.severity = error

# Override the default severity for naming styles from the 'LocalsAreCamelCase' rule
dotnet_diagnostic.IDE1006+LocalsAreCamelCase.severity = suggestion
```

**Comment by CyrusNajmabadi** [comment]
Dumb question: why ```+``` instead of ust ```.```?

**Comment by mavasani** [comment]
If we implement this, it should use a '-' as a separator, similar to category based configuration: `dotnet_analyzer_diagnostic.category-<DiagnosticCategory>.severity = <severity>`. See https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#set-rule-severity-of-multiple-analyzer-rules-at-once-in-an-editorconfig-file

**Comment by jez9999** [comment]
So how can I upvote this?

**Comment by CyrusNajmabadi** [comment]
Use the thumbs-up at the top: 

![image](https://user-images.githubusercontent.com/4564579/83180265-07a73480-a0d8-11ea-9754-b7e35216b97f.png)

**Comment by Evangelink** [comment]
I don't think I would be very valuable but @sharwell I'd be happy to help (if you need some).

**Comment by mavasani** [comment]
I believe this can be trivially implemented once #69843 goes in. We also don't need `dotnet_diagnostic` syntax for configuring such sub-severities. We can support any desired syntax that blends well with existing naming rule configuration syntax and doesn't need to use opaque IDs such as `IDE1006` in the configuration entry: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/naming-rules#general-syntax. For example, `dotnet_naming_rule.locals_are_camel_case.severity = suggestion`

**Comment by CyrusNajmabadi** [comment]
Closing out as unplanned.
